### PR TITLE
PLAT-305 - Auto increment column for redis model

### PR DESCRIPTION
**Pull request by gipsy86147**
@mflaxman -  Auto increment column for redis model

### CODE REVIEW

File: subconscious/model.py [hunk 1]
```diff
@@ -101,7 +106,7 @@ def __init__(self, allow_unknown_cols=False, **kwargs):
                     )
                 setattr(self, column.name, value)
             else:
-                assert not column.required, 'Column `{}` in `{}` is required'.format(
+                assert not (column.required and not getattr(column, 'auto', False)), 'Column `{}` in `{}` is required'.format(
```
Comment (mflaxman) [inline subconscious/model.py]: Please use exceptions and not assertions in this library. If you rebase to the latest you'll find that there aren't asserts anymore.

File: tests/test_auto_int.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+class TestUser(RedisModel):
+    id = Integer(primary_key=True, auto=True)
+    name = Column(type=str)
+    auto_id = Integer(auto=True)
+
+
+class TestAutoInt(BaseTestCase):
+
+    def test_auto(self):
```
Comment (mflaxman) [inline tests/test_auto_int.py:14]: Please break these two separate tests into two methods

File: tests/test_auto_int.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        self.loop.run_until_complete(user.save(self.db))
+        self.assertEqual(user.id, 1)
+        user = TestUser()
+        self.loop.run_until_complete(user.save(self.db))
+        self.assertEqual(user.id, 2)
+
+        # Provide id
+        user = TestUser(id=777)
+        self.loop.run_until_complete(user.save(self.db))
+        self.assertEqual(user.id, 777)
```
Comment (mflaxman) [inline tests/test_auto_int.py]: This seems like a bug, as this column also has `auto=True`. This should either = 3 or raise an exception.

File: subconscious/model.py (hunk 1 above)
Comment (mflaxman) [reply subconscious/model.py]: also, `getattr(column, 'auto', False)` is confusing

File: subconscious/column.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
+from .model import Column
```
Comment (mflaxman) [inline subconscious/column.py]: It's confusing that `Column` is in `model.py` but `Integer` is in `column.py`

File: subconscious/column.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+            composite_key=None,
+            index=None,
+            required=None,
+            enum=None,
+            sort=None, auto=True):
+        super(Integer, self).__init__(int, primary_key, composite_key, index, required, enum, sort)
+        self.auto = auto
+
+    async def auto_generate(self, db, model):
+        return await db.incr('auto:{}:{}'.format(model.key_prefix(), self.name))
```
Comment (mflaxman) [inline subconscious/column.py:71]: `auto` is ambiguous. `auto-increment`? `auto-inc`?

**Review by mflaxman — COMMENTED** [review]
more features!

File: tests/test_auto_int.py (hunk 3 above)
Comment (gipsy86147) [reply tests/test_auto_int.py]: I wanted to let the user override it.

File: subconscious/column.py (hunk 4 above)
Comment (gipsy86147) [reply subconscious/column.py]: In fact I wanted to move Column to column.py , but thought a bigger refactor in the coordinator project, may be I'll just do that

File: tests/test_auto_int.py (hunk 3 above)
Comment (mflaxman) [reply tests/test_auto_int.py]: I like that in theory, but in practice DBs usually have a unique constraint on an auto-incrementing field. The caller would probably expect the same here.

I'd be ok with an advanced feature where if you pass a flag you can over-write the auto-incrementing column, but I don't like it happening by default.

More simple would be to just throw an error and then add this field if anyone ever asks for it.

File: subconscious/column.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+class Integer(Column):
+    def __init__(
+            self,
+            primary_key=None,
+            composite_key=None,
+            index=None,
+            required=None,
+            enum=None,
+            sort=None, auto=True):
+        super(Integer, self).__init__(int, primary_key, composite_key, index, required, enum, sort)
```
Comment (mflaxman) [inline subconscious/column.py]: please use keyword arguments for this. Too easy to swap the order of two fields and make a big mistake.

File: subconscious/column.py (hunk 5 above)
Comment (gipsy86147) [reply subconscious/column.py:71]: done

File: subconscious/column.py (hunk 4 above)
Comment (gipsy86147) [reply subconscious/column.py]: done

File: subconscious/column.py (hunk 6 above)
Comment (gipsy86147) [reply subconscious/column.py]: Done

File: tests/test_auto_int.py (hunk 3 above)
Comment (gipsy86147) [reply tests/test_auto_int.py]: Done

**Review by mflaxman — APPROVED** [review]
LGTM, nice work @gipsy86147 !

Want to give @jimmysong a chance to review or just merge now?

**Comment by gipsy86147** [comment]
@jimmysong Please take a look

**Pull request merged**